interceptor/src/io/pedestal/interceptor/chain.clj

```diff
@@ -80,205 +85,215 @@
-                                                              :stage interceptor-key
-                                                              :stack new-stack})
-                                         context)
-            (::error context) (dissoc context ::queue)
-            (not= (:bindings context) pre-bindings) (assoc context ::rebind true)
-            true (recur (check-terminators context)))))))))
-
-(defn- process-all
-  [context interceptor-key]
-  ;; If we're processing leave handlers, reverse the queue
-  (let [context (if (= interceptor-key :leave) (update context ::queue reverse) context)
-        context (with-bindings (or (:bindings context)
-                                   {})
-                  (process-all-with-binding context interceptor-key))]
-    (if (::rebind context)
-      (recur (dissoc context ::rebind) interceptor-key)
-      context)))
-
-(defn- process-any-errors-with-binding
-  "Unwinds the context by invoking :error functions of Interceptors on
-  the ::stack of context, but **only** if there is an ::error present in the context."
   [context]
-  (log/debug :in 'process-any-errors :execution-id (::execution-id context))
-  (loop [context context]
-    (let [stack (::stack context)
-          execution-id (::execution-id context)]
-      (log/trace :context context)
-      (if (empty? stack)
-        context
-        (let [interceptor (peek stack)
-              pre-bindings (:bindings context)
-              old-context context
-              context (assoc context ::stack (pop stack))
-              context (if (::error context)
-                        (try-error context interceptor)
-                        context)]
-          (cond
-           (channel? context) (go-async old-context context)
-           (not= (:bindings context) pre-bindings) (assoc context ::rebind true)
-           true (recur context)))))))
-
-(defn- process-any-errors
-  "Establish the bindings present in `context` as thread local
-  bindings, and then invoke process-any-errors-with-binding.
-  Conditionally re-establish bindings if a change in bindings is made by an
-  interceptor."
+  (doseq [enter-async-fn (::enter-async context)]
+    (enter-async-fn context))
+  ;; Only call it once per execution:
+  (dissoc context ::enter-async))
+
+(declare ^:private invoke-interceptors-binder)
+
+(defn- handle-async
+  "Invoked when an interceptor returns a channel, rather than a context map.
+
+  context-ch: will convey context
+  execution-context: context passed to the active interceptor
+  active-interceptor: interceptor invoked
+  continuation: the continuation map
+  "
+  [context-ch execution-context active-interceptor continuation]
+  (log/debug :in 'handle-async
+             :execution-id (::execution-id execution-context)
+             :interceptor (name active-interceptor))
+  (let [context' (prepare-for-async execution-context)]
+    (async/take! context-ch
+                 (fn [new-context]
```

Comment:
This path can lead to :entrer-async callbacks being invoked again.  Needs to dissoc :enter-async.